### PR TITLE
Add params for solver callback

### PR DESCRIPTION
**Pull request by sunce86**
Follow up PR for https://github.com/cowprotocol/services/pull/661

Defines struct `TransactionWithMessage` with data to be sent to solvers in case of simulation failure. 

Transaction input data + block number represents the bare minimum needed for simulation reconstruction.
~All other data is more or less static or irrelevant (`from`, `to`, `gas`, `gas_price`) therefore being skipped.~
Edit: added `from` and `to`.

### CODE REVIEW

File: crates/solver/src/settlement_ranker.rs [hunk 1]
```diff
@@ -159,12 +162,15 @@ impl SettlementRanker {
...
             errors.len(),
         );
-        for (solver, settlement, _, error) in &errors {
+        for (solver, settlement, _, block_number, error) in &errors {
             solver.notify_auction_result(
                 auction_id,
                 AuctionResult::Rejected(SolverRejectionReason::SimulationFailure(
-                    error.to_string(),
-                    call_data(settlement.clone().into()),
+                    SimulationFailureParams {
```
Comment (poolpitako) [inline crates/solver/src/settlement_ranker.rs:169]: shouldn't you add the settlement contract address just in case it changes?

**Review by ennmichael — APPROVED** [review]
LGTM

File: crates/solver/src/settlement_ranker.rs (hunk 1 above)
Comment (poolpitako) [reply crates/solver/src/settlement_ranker.rs:169]: just for completeness.
It's kind of weird solvers gets the `call_data` but not the contract which it is supposed to send to.

Comment (sunce86) [reply crates/solver/src/settlement_ranker.rs:176]: Added `from` and `to` fields, with a bit of refactoring to make it easier to add additional fields in the future (if needed).

File: crates/solver/src/driver_logger.rs [hunk 2]
```diff
@@ -147,13 +147,24 @@ impl DriverLogger {
...
+                                 transaction,
+                                 ..
+                             },
+                         ..
+                     }| {
+                        (
+                            solver.account().clone(),
+                            settlement.clone(),
+                            transaction.access_list.clone(),
+                        )
```
Comment (MartinquaXD) [inline crates/solver/src/driver_logger.rs:165]: style nit: The destructuring looks quite noisy here. What do you think about something like 👇  instead?
The same idea could also be applied to the destructuring in the for loop.

```suggestion
                    |simulation_with_error| {
                        let simulation = &simulation_with_error.simulation;
                        (
                            simulation.solver.account().clone(),
                            simulation.settlement.clone(),
                            simulation.transaction.access_list.clone(),
                        )
                    }
```

File: crates/solver/src/settlement_rater.rs [hunk 3]
```diff
@@ -19,14 +19,8 @@ use web3::types::AccessList;
...
 pub type RatedSolverSettlement = (Arc<dyn Solver>, RatedSettlement, Option<AccessList>);
 
-pub struct SimulationDetails {
-    pub settlement: Settlement,
-    pub solver: Arc<dyn Solver>,
-    /// Which storage the settlement tries to access. Contains `None` if some error happened while
-    /// estimating the access list.
-    pub access_list: Option<AccessList>,
-    /// The outcome of the simulation. Contains either how much gas the settlement used or the
-    /// reason why the transaction reverted during the simulation.
```
Comment (MartinquaXD) [inline crates/solver/src/settlement_rater.rs:29]: IMO without this comment it's not obvious how the field is to be understood.

File: crates/solver/src/solver.rs [hunk 4]
```diff
@@ -97,6 +97,13 @@ pub enum AuctionResult {
     Rejected(SolverRejectionReason),
 }
 
+pub struct TransactionWithMessage {
+    /// Data needed to reconstruct the simulation
```
Comment (MartinquaXD) [inline crates/solver/src/solver.rs:101]: nit: I guess this comment should rather on the `SimulatedTransaction` struct itself.

File: crates/solver/src/solver.rs [hunk 5]
```diff
@@ -186,13 +192,29 @@ pub type Solvers = Vec<Arc<dyn Solver>>;
...
+    /// Which storage the settlement tries to access. Contains `None` if some error happened while
+    /// estimating the access list.
+    pub access_list: Option<AccessList>,
+    /// Solver address
+    pub from: H160,
+    /// GPv2 settlement contract address
+    pub to: H160,
+    /// Transaction input data
+    pub data: Vec<u8>,
+}
```
Comment (MartinquaXD) [inline crates/solver/src/solver.rs:208]: nit: `\n` after the structs.

**Review by MartinquaXD — APPROVED** [review]
Looks good overall. Just a few style nits/suggestions.

File: crates/solver/src/driver_logger.rs (hunk 2 above)
Comment (sunce86) [reply crates/solver/src/driver_logger.rs]: Yes, I've abused a bit the pattern matching here :) Fixed.

File: crates/solver/src/settlement_rater.rs (hunk 3 above)
Comment (sunce86) [reply crates/solver/src/settlement_rater.rs]: Reverted.

File: crates/solver/src/solver.rs (hunk 4 above)
Comment (sunce86) [reply crates/solver/src/solver.rs]: Moved.

File: crates/solver/src/solver.rs (hunk 5 above)
Comment (sunce86) [reply crates/solver/src/solver.rs:210]: Added.

**Pull request merged**